hazelcast/src/main/java/com/hazelcast/quorum/QuorumServiceImpl.java

```diff
@@ -0,0 +1,290 @@
+            for (QuorumListenerConfig listenerConfig : config.getListenerConfigs()) {
+                if (listenerConfig != null) {
+                    QuorumListener listener = null;
+                    if (listenerConfig.getImplementation() != null) {
+                        listener = listenerConfig.getImplementation();
+                    } else if (listenerConfig.getClassName() != null) {
+                        try {
+                            listener = ClassLoaderUtil
+                                    .newInstance(nodeEngine.getConfigClassLoader(), listenerConfig.getClassName());
+                        } catch (Exception e) {
+                            throw ExceptionUtil.rethrow(e);
+                        }
+                    }
+                    if (listener != null) {
+                        String instanceName = configEntry.getKey();
+                        addQuorumListener(instanceName, listener);
+                    }
+                }
+            }
+        }
+    }
+
+    public void addQuorumListener(String name, QuorumListener listener) {
+        eventService.registerLocalListener(SERVICE_NAME, name, listener);
+    }
+
+
+    public void resolveQuorum(Operation op) {
+        if (isNamedOperation(op) && hasQuorumAwareService(op)) {
+            String quorumName = getQuorumName(op);
+            if (quorumName != null && isQuorumNeeded(quorumName, op)) {
+                doQuorum(quorumName);
+            }
+        }
+    }
+
+    private String getQuorumName(Operation op) {
+        QuorumAwareService service = op.getService();
+        String name = ((NamedOperation) op).getName();
+        return service.getQuorumName(name);
+    }
+
+    private boolean hasQuorumAwareService(Operation op) {
+        return op.getService() instanceof QuorumAwareService;
+    }
+
+    private boolean isNamedOperation(Operation op) {
+        return op instanceof NamedOperation;
+    }
+
+    private boolean isReadOperation(Operation op) {
+        return op instanceof ReadonlyOperation;
+    }
+
+    private boolean isWriteOperation(Operation op) {
+        return op instanceof MutatingOperation;
+    }
+
+    //CHECKSTYLE:OFF
+    private boolean isQuorumNeeded(String quorumName, Operation op) {
+        QuorumConfig quorumConfig = quorumConfigs.get(quorumName);
+        QuorumType type = quorumConfig.getType();
+        switch (type) {
+            case WRITE:
+                if (!isWriteOperation(op)) {
+                    return false;
+                }
+                return true;
+            case READ:
+                if (!isReadOperation(op)) {
+                    return false;
+                }
+                return true;
+            case READ_WRITE:
+                if (!isReadOperation(op) && !isWriteOperation(op)) {
+                    return false;
+                }
+                return true;
+            default:
+                return false;
```

Comment:
Default ---> throw error. This is a violation of the fail fast, fail hard principle.